### PR TITLE
[tune] Document trainable attributes and enable user-checkpointing

### LINKED ISSUE #4363: [tune] Better documentation for 'done'
It is not clear in the docs where `done` is used (to terminate trials from within the trial).

### PR DESCRIPTION
**Pull request by richardliaw**
## What do these changes do?

Documents the trainable attributes. 

## Related issue number

Closes #4363

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

File: python/ray/rllib/agents/mock.py [hunk 1]
```diff
@@ -32,11 +34,15 @@ def _train(self):
...
             raise Exception("mock error")
-        return dict(
+        result = dict(
             episode_reward_mean=10,
             episode_len_mean=10,
             timesteps_this_iter=10,
             info={})
+        if self.config["user_checkpoint_freq"] > 0 and self.iteration > 0:
+            if self.iteration % self.config["user_checkpoint_freq"] == 0:
+                result.update({tune_result.SHOULD_CHECKPOINT: True})
```
Comment (hartikainen) [inline python/ray/rllib/agents/mock.py]: What if user has explicitly said `tune_result.SHOULD_CHECKPOINT: False`? Should this be 
```suggestion
                result.setdefault(tune_result.SHOULD_CHECKPOINT, True)
```

File: python/ray/tune/trainable.py [hunk 2]
```diff
@@ -395,11 +403,45 @@ def stop(self):
...
+    def iteration(self):
+        """Current training iteration.
+
+        This value is automatically incremented every time `train()` is called
+        and is automatically inserted into the training result dict.
+
+        """
+        return self._iteration
+
+    def get_config(self):
```
Comment (hartikainen) [inline python/ray/tune/trainable.py]: To be consistent with the rest of the methods above, should could be, although I don't know if it's worth changing in this PR.
```
@property
def config(self):
        """Returns configuration passed in by Tune."""
        return self._config
```

**Review by hartikainen — APPROVED** [review]
Looks good to me.

File: python/ray/rllib/agents/mock.py (hunk 1 above)
Comment (richardliaw) [reply python/ray/rllib/agents/mock.py]: Oh, this is only for testing purposes and isn't extended.

File: python/ray/tune/trainable.py (hunk 2 above)
Comment (richardliaw) [reply python/ray/tune/trainable.py]: This ran into issues with RLlib here (`trainer.py`):

```python
        # Merge the supplied config with the class default
        merged_config = copy.deepcopy(self._default_config)
        merged_config = deep_update(merged_config, config,
                                    self._allow_unknown_configs,
                                    self._allow_unknown_subkeys)
        self.raw_user_config = config
        self.config = merged_config
        Trainer._validate_config(self.config)
        if self.config.get("log_level"):
            logging.getLogger("ray.rllib").setLevel(self.config["log_level"])
```
I'll merge first and we can see later if we need to fix (also @ericl doesn't like properties :) )

File: python/ray/rllib/agents/mock.py (hunk 1 above)
Comment (hartikainen) [reply python/ray/rllib/agents/mock.py]: Lol fo course, my bad

**Pull request merged**